### DISCUSSION TITLE
Unable to Install Multiple Versions of SCons Side-by-Side for SCONS 4.2.0

### DISCUSSION THREAD

**Discussion by BGrandhi** [SCons/scons#4258] [Q&A]
Hi,
For my development works, I need to have multiple versions of scons installed on my PC (SCONS-3.1.2 and SCONS-4.2.0). I went through the SCONS user guide for finding some info for a way to achieve this and find out I can use an command **python setup.py install --version-lib** 
The above command worked fine when I am trying to install SCONS 3.1.2 but when I try to install SCONS 4.2.0 then it fails saying **error: option --version-lib not recognized** . 
Can you please help me with a way to achieve this so that I can work with two versions of scons on my PC? 
Thanks and please let me know if you need any more info.

**Comment by BGrandhi** [comment]
Here is the cmd output I am receiving  when I try to install SCONS 4.2.0

C:\Users\grb1rlz\Desktop\AllStuff\EXE\AllOrganized\BuildTools\scons\scons-4.2.0>python setup.py install --version-lib
C:\Users\GRB1RLZ\AppData\Local\Programs\Python\Python39\lib\site-packages\setuptools\dist.py:642: UserWarning: Usage of dash-separated 'project-urls' will not be supported in future versions. Please use the underscore name 'project_urls' instead
  warnings.warn(
C:\Users\GRB1RLZ\AppData\Local\Programs\Python\Python39\lib\site-packages\setuptools\dist.py:642: UserWarning: Usage of dash-separated 'dist-dir' will not be supported in future versions. Please use the underscore name 'dist_dir' instead
  warnings.warn(
usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
   or: setup.py --help [cmd1 cmd2 ...]
   or: setup.py --help-commands
   or: setup.py cmd --help

error: option --version-lib not recognized

**Comment by bdbaddog** [comment]
This is not a bug. The proper channel to ask is via the users mailing list, the IRC channel or the discord server.
https://scons.org/contact.html

You should use python virtualenvs.
https://github.com/SCons/scons/blob/master/README.rst#installation

**Comment by bdbaddog** [comment]
Moved to discussion.

Where did you see instructions to use --version-lib? That's obsolete information.

**Reply by BGrandhi** [reply]
Thanks for the response. I have seen it in the user guide section **1.3.1. Building and Installing Multiple Versions of SCons Side-by-Side** in the below link.
https://scons.org/doc/4.2.0/HTML/scons-user.html#idp140211710046352

**Reply by bdbaddog** [reply]
Gotcha. Ok We'll make a note to update that section.

**Reply by bdbaddog** [reply]
Were you able to setup the virtualenvs?

**Reply by BGrandhi** [reply]
Yeah I just tried that based on the link you have provided. Now I can use both versions of SCONS on my PC without issues. Thank you so much for the quick help. I really appreciate it.

**Comment by bdbaddog** [comment]
Added bug to remove/update that obsoleted section of users guide: https://github.com/SCons/scons/issues/4259

**Comment by mwichmann** [comment]
There's a simpleminded virtualenv recipe in the cookbook: https://scons-cookbook.readthedocs.io/en/latest/#setting-up-a-python-virtualenv-for-scons

How you do this will depend on whether you're able to use the same Python for both versions.  If you can, make two virtualenvs with different names, and install SCons once in each one, something like this:
```console
cd oldscons
source bin/activate
pip install scons==3.1.2
deactivate
cd ../newscons
source bin/activate
pip install scons==4.2.0
```
(the latter assuming you need 4.2.0, if you leave out the version specifier  you'll get the latest, which is 4.4.0)

Now whichever virtualenv you have activated, you'll get the SCons in that env - you're clearly not limited to two of them.  This is the preferred Python way of keeping isolated setups that don't interfere with one another (or with the system Python installation).

The above was Linux syntax, but the recipe has tabs for Mac and Windows, so you can see what those might look like if needed.

**Reply by BGrandhi** [reply]
Thanks for the info. As of now, I am able to install multiple versions of scons using virtual environments  though downloading the scons installation folders . But if I use PIP then I am getting connection failed. I think this may be because of the firewall from my organization. 
Please let me know if I am doing something wrong. Here is the error I am getting when I use PIP

(env_latest) C:\WINDOWS\system32>pip list --outdated
WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000001D5CC97CB20>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/pip/
WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000001D5CC97CD90>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/setuptools/
WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000001D5CC9A7310>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/setuptools/
WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000001D5CC9A7130>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/pip/
WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000001D5CC9A7550>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/setuptools/
WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000001D5CC9A7730>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/pip/
WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000001D5CC9A7B50>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/pip/
WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000001D5CC9A7970>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/setuptools/
WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000001D5CC9A7D90>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/pip/
WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x000001D5CC9A7F70>: Failed to establish a new connection: [Errno 11001] getaddrinfo failed')': /simple/setuptools/

**Comment by mwichmann** [comment]
pip can take a `--proxy` argument, if you know what your organization's proxy setup is.

**Reply by BGrandhi** [reply]
Thank you, I will find the proxy info internally
